libbpf-rs/src/object.rs

```diff
@@ -1,84 +1,353 @@
+            let c_name = util::str_to_cstring(name.as_ref())?;
+            let ptr =
+                unsafe { libbpf_sys::bpf_object__find_program_by_name(self.ptr, c_name.as_ptr()) };
+            if ptr.is_null() {
+                Ok(None)
+            } else {
+                let owned_name = name.as_ref().to_owned();
+                self.progs.insert(owned_name, ProgramBuilder::new(ptr));
+                Ok(self.progs.get_mut(name.as_ref()))
+            }
+        }
     }
+}
 
-    pub fn prog<T: AsRef<str>>(&mut self, _name: T) -> Option<&mut ProgramBuilder> {
-        unimplemented!();
+impl Drop for Object {
+    fn drop(&mut self) {
+        unsafe {
+            libbpf_sys::bpf_object__close(self.ptr);
+        }
     }
 }
 
 /// Represents a parsed but not yet loaded map.
 ///
 /// Some methods require working with raw bytes. You may find libraries such as
 /// [`plain`](https://crates.io/crates/plain) helpful.
-pub struct MapBuilder {}
+pub struct MapBuilder {
+    ptr: *mut libbpf_sys::bpf_map,
+    name: String,
+    attrs: libbpf_sys::bpf_create_map_attr,
+    initial_val: Option<Vec<u8>>,
+}
 
 impl MapBuilder {
-    pub fn set_map_ifindex(&mut self, _idx: u32) -> &mut Self {
-        unimplemented!();
+    fn new(ptr: *mut libbpf_sys::bpf_map, name: String, btf_fd: i32) -> Self {
+        // bpf_map__def can return null but only if it's passed a null. Object::map
+        // already error checks that condition for us.
+        let def = unsafe { ptr::read(libbpf_sys::bpf_map__def(ptr)) };
+
+        let mut attrs = libbpf_sys::bpf_create_map_attr::default();
+        attrs.map_type = def.type_;
+        attrs.key_size = def.key_size;
+        attrs.value_size = def.value_size;
+        attrs.max_entries = def.max_entries;
+        attrs.map_flags = def.map_flags;
+
+        if attrs.map_type == (MapType::PerfEventArray as u32) && attrs.max_entries == 0 {
+            attrs.max_entries = unsafe { libbpf_sys::libbpf_num_possible_cpus() as u32 };
+        }
+
+        if btf_fd >= 0 {
+            attrs.btf_fd = btf_fd as u32;
+            attrs.btf_value_type_id = unsafe { libbpf_sys::bpf_map__btf_value_type_id(ptr) };
+            attrs.btf_key_type_id = unsafe { libbpf_sys::bpf_map__btf_key_type_id(ptr) };
+        }
+
+        MapBuilder {
+            ptr,
+            attrs,
+            name,
+            initial_val: None,
+        }
+    }
+
+    pub fn set_map_ifindex(&mut self, idx: u32) -> &mut Self {
+        self.attrs.map_ifindex = idx;
+        self
     }
 
-    pub fn set_max_entries(&mut self, _entries: u32) -> &mut Self {
-        unimplemented!();
+    pub fn set_max_entries(&mut self, entries: u32) -> &mut Self {
+        self.attrs.max_entries = entries;
+        self
     }
 
-    pub fn set_initial_value(&mut self, _data: &[u8]) -> &mut Self {
-        unimplemented!();
+    pub fn set_initial_value(&mut self, data: &[u8]) -> &mut Self {
+        self.initial_val = Some(data.to_vec());
```

Comment:
why not just calling `libbpf_sys::bpf_map__set_initial_value` here?

Reply:
Yeah, I guess that'll work b/c the final `load()` (will probably rename to `build()`) acts as a barrier so users won't ever get access to a partially built map

Reply:
Hmm, actually the reason it's like that is b/c bpf_map__set_initial_value() can fail and it'd be weird to have a builder option be fallible